client/allocrunner/csi_hook_test.go

```diff
@@ -0,0 +1,253 @@
+package allocrunner
+
+import (
+	"context"
+	"fmt"
+	"path/filepath"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/hashicorp/nomad/client/allocrunner/interfaces"
+	"github.com/hashicorp/nomad/client/pluginmanager"
+	"github.com/hashicorp/nomad/client/pluginmanager/csimanager"
+	cstructs "github.com/hashicorp/nomad/client/structs"
+	"github.com/hashicorp/nomad/helper/testlog"
+	"github.com/hashicorp/nomad/nomad/mock"
+	"github.com/hashicorp/nomad/nomad/structs"
+	"github.com/hashicorp/nomad/plugins/drivers"
+)
+
+var _ interfaces.RunnerPrerunHook = (*csiHook)(nil)
+var _ interfaces.RunnerPostrunHook = (*csiHook)(nil)
+
+// TODO https://github.com/hashicorp/nomad/issues/11786
+// we should implement Update as well
+// var _ interfaces.RunnerUpdateHook = (*csiHook)(nil)
+
+func TestCSIHook(t *testing.T) {
+
+	alloc := mock.Alloc()
+	logger := testlog.HCLogger(t)
+
+	testcases := []struct {
+		name                   string
+		volumeRequests         map[string]*structs.VolumeRequest
+		expectedMounts         map[string]*csimanager.MountInfo
+		expectedMountCalls     int
+		expectedUnmountCalls   int
+		expectedClaimCalls     int
+		expectedUnpublishCalls int
+	}{
+
+		{
+			name: "simple case",
+			volumeRequests: map[string]*structs.VolumeRequest{
+				"vol0": {
+					Name:           "vol0",
+					Type:           structs.VolumeTypeCSI,
+					Source:         "testvolume0",
+					ReadOnly:       true,
+					AccessMode:     structs.CSIVolumeAccessModeSingleNodeReader,
+					AttachmentMode: structs.CSIVolumeAttachmentModeFilesystem,
+					MountOptions:   &structs.CSIMountOptions{},
+					PerAlloc:       false,
+				},
+			},
+			expectedMounts: map[string]*csimanager.MountInfo{
+				"vol0": &csimanager.MountInfo{Source: fmt.Sprintf(
+					"test-alloc-dir/%s/testvolume0/ro-file-system-single-node-reader-only", alloc.ID)},
+			},
+			expectedMountCalls:     1,
+			expectedUnmountCalls:   0, // not until this is done client-side
+			expectedClaimCalls:     1,
+			expectedUnpublishCalls: 1,
+		},
+
+		{
+			name: "per-alloc case",
+			volumeRequests: map[string]*structs.VolumeRequest{
+				"vol0": {
+					Name:           "vol0",
+					Type:           structs.VolumeTypeCSI,
+					Source:         "testvolume",
```

Comment:
```suggestion
					Source:         "testvolume0",
```
I'm not sure if this was intentional, and also doesn't seem to matter apart from consistency with the other tests.

Reply:
It was intentional when I thought https://github.com/hashicorp/nomad/issues/11787 was a bug, but I forgot to back it out when I closed that issue. Will fix.